### ISSUE TITLE
PingFederate: does not seem to respect LICENSE_DIR

### ISSUE THREAD

**Issue by coryasilva** [pingidentity/pingidentity-devops-getting-started#180]
**Describe the bug**
I am setting the LICENSE_DIR but the pingfederate does not seem to use it.

**To Reproduce**
1. set LICENSE_DIR to a nondefault value
2. Run docker container
3. View logs

**Expected behavior**
I would expect that the license file directory would be expected or at least copied to `/opt/out/instance/server/default/conf/`

**Logs (redacted some lines for brevity)**
```sh
----- Starting hook: /opt/staging/hooks/03-build-run-plan.sh
Missing SERVER_ROOT: /opt/out/instance
################################################################################
#    Docker Image Information
################################################################################
                     IMAGE_VERSION : pingfederate-alpine-az11-10.2.0-210112-4797
                     IMAGE_GIT_REV : 4797657ce77b559f4576a2df671b3e818b5a9057
################################################################################
#    Directory Variables
################################################################################
                       LICENSE_DIR : /blah
################################################################################
#    License Key Info
################################################################################
                 LICENSE_FILE_NAME : pingfederate.lic
                LICENSE_SHORT_NAME : PF
                   LICENSE_VERSION : 10.2
         MUTE_LICENSE_VERIFICATION : --- empty ---

----- Starting hook: /opt/staging/hooks/17-check-license.sh
Verifying license with a network query to https://license.pingidentity.com.
You may opt out of this setting environment variable 'MUTE_LICENSE_VERIFICATION=yes'.
   License File: /blah/pingfederate.lic
        License: 00131153
----- Starting hook: /opt/staging/hooks/80-post-start.sh
INFO: waiting for healthy admin before post-start..
2021-01-29 22:43:44,646  WARN  [org.sourceid.util.license.PingLicense]
**********************************************************
No license found for this server (/opt/out/instance/server/default/conf/pingfederate.lic)
You do not have a valid license configured. To obtain a valid license, please contact your vendor's Customer Support.
**********************************************************
----- Starting hook: /opt/staging/hooks/83-configure-admin.sh
PingFederate running...
INFO: new server found, must create admin
**********************************************************
No license file was found.
You do not have a valid license configured. To obtain a valid license, please contact your vendor's Customer Support.
**********************************************************
{
  "resultId": "invalid_license",
  "message": "The license for this PingFederate instance is missing or invalid."
}
error attempting to create admin
CONTAINER FAILURE: Error running 83-configure-admin.sh
```

**Environment:**
AWS, EKS, pingfederate:2012-alpine-10.2.0

**Comment by kenneth-ping** [comment]
Jira https://jira.pingidentity.com/browse/GDO-682 filed for this issue.

**Comment by coryasilva** [comment]
### Temp Workaround

`/opt/in/hooks/50-before-post-start.sh.pre`

```sh
#!/usr/bin/env sh
set -x
# https://github.com/pingidentity/pingidentity-devops-getting-started/issues/180
echo "Hack for GDO-682: Copy License to /opt/out/instance/server/default/conf/"
mkdir -pv /opt/out/instance/server/default/conf/
cp -p "${LICENSE_DIR}/${LICENSE_FILE_NAME}" "/opt/out/instance/server/default/conf/${LICENSE_FILE_NAME}"
```

**Comment by samir-gandhi** [comment]
Hey @coryasilva, 
right, this is one way to do it. 
but, what's the reason for not using the default license dir? 
As background, the `$LICENSE_DIR/$LICENSE_FILE_NAME` variable is really just to point to a location where the _container_ will validate the license, not the product. This purpose is to help the container know if it needs to pull an eval license. 

so the concept is:
1. container starts, looks for `$LICENSE_DIR/$LICENSE_FILE_NAME` validates that it exists and is not expired. 
2. during startup, if /opt/out doesn't exist /opt/staging is copied over. this is is how a regular license would make it to the location where PF looks for it. 
3. PF starts and looks for the license. 

I am thinking, you may have had trouble getting the license to the right spot because of step 2? since /opt/out existed it didn't copy over the license? I don't know if this should need overwriting the variables though

**Comment by coryasilva** [comment]
@samir-gandhi I am passing the license in through a k8s secret volume mount. The default secrets directory is `/run/secrets/` which is where kubernetes mounts its [service account creds](https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/#serviceaccount-admission-controller) 
so I am using a subfolder to stop the secret mount from wiping out the SA token.

The reason I am not simply mounting the secret to `/opt/in` is because that is a docker layer and I cannot mount a (readonly) secret volume to a docker layer.  [There is some esoteric reason that I half understand...](https://github.com/kubernetes/kubernetes/pull/58720)

**Comment by samir-gandhi** [comment]
Got it, thanks for the additional context.

**Comment by arnaudlacour** [comment]
We probably don't document this well enough but the `LICENSE_DIR` environment variable is not meant to tell the server where to pick up a license file. Rather, it lets our container automation know _where to expect the license_ so it can check the license is present before the server startup is even attempted. It is also used by so that we can drop a temporary license obtained against the devops license service in the correct location for each of our products.
In short, it's not really up for user modification.

**Comment by coryasilva** [comment]
@arnaudlacour While I understand your stance on that, it is a little incongruent with the PingData containers which do respect the configuration of `LICENSE_DIR`.

On a slight tangent, you all should reconsider validating the license in such a way; it seems like a mixed concern.  Why does my OS (container) care about my application license?

And while on the topic of unsolicited feedback; I also think these containers would benefit from an approach that valued [composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance).

I appreciate all the work you and your team have done here and thank you all for being so responsive!

**Comment by arnaudlacour** [comment]
The container checks for the license to _fail fast_.
The idea is to be able to wipe a container very quickly for anything obviously untoward.
Scaffolding a whole JVM up only to realize after the process is up that we can't process any requests because of a missing license file seems like an easy optimization.

100 on Composition over Inheritance.
We're moving in that direction in a "docker release near you" later this year. We refer to these as Mark II images and they are in the design stage at the moment.

thanks for the feedback, solicited or not, we're here to make it incremental easier/faster/better for **you** ultimately.

**Comment by tsigle** [comment]
@coryasilva , your comments are well founded and taken with respect to the PingData containers respecting the location of the LICENSE_DIR, but PF/PA not.  Unfortunately, those products require a very specific location for that license to be present in.  I don't see that requirement changing in the near future, from what I can tell.

**Comment by coryasilva** [comment]
@arnaudlacour Failing fast makes sense; I had not considered that.  Also, looking forward to the mark II images.

@tsigle Fair enough, I posted my silly hack in case people run into it.  How are other people accomplishing this in kubernetes given the constraints I explained [above](https://github.com/pingidentity/pingidentity-devops-getting-started/issues/180#issuecomment-771241487)?
